File: 210.course_search2/findOrder.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
    var graph = [];
    for(let i = 0; i < numCourses; i++){
        graph[i] = [];
    }
    prerequisites.forEach(function(x){
        graph[x[1]].push(x[0]);
    });
    //initialized graph and visited arrays
    var visited = [];
    for(let i = 0; i < numCourses; i++){
        visited[i] = 0;
    }

    var dfs = function(i,graph,visited,record){
        //if has been visited
        if(visited[i] == 1){
            return true;
        }
        //if is being visited
        if(visited[i] == -1){
            return false;
        }
        visited[i] = -1;

        for(let j = 0; j < graph[i].length; j++){
            if(!dfs(graph[i][j],graph,visited,record)){
                return false;
            }
        }
        record.push(i);
        visited[i] = 1;
        return true;

    };
    var record = [];
    for(let i = 0; i < numCourses; i++){
        if(!dfs(i,graph,visited,record)){
            return [];
        }
    }
    return record.reverse();

};

File: 210.course_search2/findOrder.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        def dfs(i, visited, graph, ret):
            #Stop visiting checked nodes
            if visited[i] == 1:
                return True
            #If cycle is detected, return False
            if visited[i] == -1:
                return False

            visited[i] = -1
            for n in graph[i]:
                if not dfs(n, visited, graph, ret):
                    return False
            #collecting results
            ret.append(i)

            visited[i] = 1
            return True

        #Initializing graph data
        visited = [0] * numCourses
        graph = {x:[] for x in xrange(numCourses)}
        for p in prerequisites:
            graph[p[1]].append(p[0])

        #collecting topological order
        ret = []
        for i in xrange(numCourses):
            if not dfs(i, visited, graph, ret):
                return []

        return ret[::-1]